ws.c

```diff
@@ -0,0 +1,433 @@
+{
+	if (evbuffer_get_length(bufferevent_get_output(bev)) == 0) {
+		evws_connection_free(ctx);
+	}
+}
+
+static void
+close_event_cb(struct bufferevent *bev, short what, void *ctx)
+{
+	evws_connection_free(ctx);
+}
+
+void
+evws_close(struct evws_connection *evws, uint16_t reason)
+{
+	uint8_t fr[4] = {0x8 | 0x80, 2, 0};
+	struct evbuffer *output;
+	uint16_t *u16;
+
+	if (evws->closed)
+		return;
+	evws->closed = true;
+
+	u16 = (uint16_t *)&fr[2];
+	*u16 = htons((int16_t)reason);
+	output = bufferevent_get_output(evws->bufev);
+	evbuffer_add(output, fr, 4);
+
+	/* wait for close frame writing complete and close connection */
+	bufferevent_setcb(
+		evws->bufev, NULL, close_after_write_cb, close_event_cb, evws);
+}
+
+static void
+evws_force_disconnect_(struct evws_connection *evws)
+{
+	evws_close(evws, WS_CR_NONE);
+}
+
+/* parse base frame according to
+ * https://www.rfc-editor.org/rfc/rfc6455#section-5.2
+ */
+static enum WebSocketFrameType
+get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
+	int *out_len)
+{
+	unsigned char opcode;
+	unsigned char fin;
+	unsigned char masked;
+	int payload_len;
+	int pos;
+	int length_field;
+	unsigned int mask;
+
+	if (buf_len < 2) {
+		return INCOMPLETE_DATA;
+	}
+
+	opcode = in_buffer[0] & 0x0F;
+	fin = (in_buffer[0] >> 7) & 0x01;
+	masked = (in_buffer[1] >> 7) & 0x01;
+
+	payload_len = 0;
+	pos = 2;
+	length_field = in_buffer[1] & (~0x80);
+
+	if (length_field <= 125) {
+		payload_len = length_field;
+	} else if (length_field == 126) { /* msglen is 16bit */
+		if (buf_len < 4)
+			return INCOMPLETE_DATA;
+		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		pos += 2;
+	} else if (length_field == 127) { /* msglen is 64bit */
+		if (buf_len < 10)
+			return INCOMPLETE_DATA;
+		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
+		pos += 8;
+	}
+	if (buf_len < payload_len + pos + (masked ? 4 : 0)) {
+		return INCOMPLETE_DATA;
+	}
+
+	/* According to RFC it seems that unmasked data should be prohibited
+	 * but we support it for nonconformant clients
+	 */
+	if (masked) {
+		unsigned char *c;
+
+		mask = *((unsigned int *)(in_buffer + pos));
+		pos += 4;
+
+		/* unmask data */
+		c = in_buffer + pos;
+		for (int i = 0; i < payload_len; i++) {
+			c[i] = c[i] ^ ((unsigned char *)(&mask))[i % 4];
+		}
+	}
+
+	*payload_ptr = in_buffer + pos;
+	*out_len = payload_len;
+
+	/* are reserved for further frames */
+	if ((opcode >= 3 && opcode <= 7) || (opcode >= 0xb))
+		return ERROR_FRAME;
+
+	if (opcode <= 0x3 && !fin) {
+		return INCOMPLETE_FRAME;
+	}
+	return opcode;
+}
+
+
+static void
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len, header_sz;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
```

Comment:
May fail

Reply:
Do you mean fail of `evbuffer_pullup` (I know that it may return NULL, but in the previous step we determine `in_len` and pullup exactly `in_len` size)?

Reply:
Yes, failure of `evbuffer_pullup()`. It may be possible because of pinned chains for example, this should not happens here, but it should be definitely handled anyway.